src/compiler/checker.ts

```diff
@@ -22983,10 +22983,10 @@ namespace ts {
             return isLengthPushOrUnshift || isElementAssignment;
         }
 
-        function isDeclarationWithExplicitTypeAnnotation(declaration: Declaration) {
-            return (declaration.kind === SyntaxKind.VariableDeclaration || declaration.kind === SyntaxKind.Parameter ||
-                declaration.kind === SyntaxKind.PropertyDeclaration || declaration.kind === SyntaxKind.PropertySignature) &&
-                !!getEffectiveTypeAnnotationNode(declaration as VariableDeclaration | ParameterDeclaration | PropertyDeclaration | PropertySignature);
+        function isDeclarationWithExplicitTypeAnnotation(node: Declaration) {
+            return (isVariableDeclaration(node) || isPropertyDeclaration(node) || isPropertySignature(node) || isParameter(node)) &&
+                !!(getEffectiveTypeAnnotationNode(node) ||
+                    isInJSFile(node) && hasInitializer(node) && node.initializer && isFunctionExpressionOrArrowFunction(node.initializer) && getEffectiveReturnTypeNode(node.initializer));

```

Comment:
`getEffectiveReturnTypeNode` already does the correct check for JS files, and I don't think this needs to be restricted to function/arrow expressions:
```suggestion
                    hasInitializer(node) && node.initializer && getEffectiveReturnTypeNode(node.initializer));
```

Reply:
@sandersn Thanks. It seems `isInJSFile` is useless. I used `isFunctionExpressionOrArrowFunction` to follow a type signature that is compatible with `getEffectiveReturnTypeNode` in order to avoid type casting. Can `node.initializer` contain more than just function/arrow expressions?

Reply:
One note about `isInJSFile`. Is it valid to allow the following case? 
```ts
const assert = (value: unknown): asserts value => {}
assert(typeof x === "string"); 
```

TS throws an error for that case.
```
error TS2775: Assertions require every name in the call target to be declared with an explicit type annotation.```

Reply:
Looking at [the original PR](https://github.com/microsoft/TypeScript/pull/32695):

> This particular rule exists so that control flow analysis of potential assertion calls doesn't circularly trigger further analysis.)

So I think neither the TS nor the equivalent JS is safe:

```js
const assert = /** @param {unknown} value @returns {asserts value} */(value) => {}
```

Reply:
@sandersn Thanks